fs/nfs/rpc_clnt.c

```diff
@@ -297,29 +299,47 @@ static int rpcclnt_socket(FAR struct rpcclnt *rpc, in_port_t rport)
 static int rpcclnt_send(FAR struct rpcclnt *rpc,
                         FAR void *call, int reqlen)
 {
+  FAR uint8_t *pkt = NULL;
   uint32_t mark;
   int ret = OK;
 
-  /* Send the record marking(RM) for stream only */
+  /* Prepare the record marking(RM) for stream only
+   * NOTE: Sending a separate packet does not work with Linux host
+   */
 
   if (rpc->rc_sotype == SOCK_STREAM)
     {
-      mark = txdr_unsigned(0x80000000 | reqlen);
-      ret = psock_send(&rpc->rc_so, &mark, sizeof(mark), 0);
-      if (ret < 0)
+      pkt = (uint8_t *)kmm_malloc(sizeof(mark) + reqlen);
+
+      if (NULL == pkt)
         {
           ferr("ERROR: psock_send mark failed: %d\n", ret);
-          return ret;
+          return -ENOMEM;
         }
+
+      mark = txdr_unsigned(0x80000000 | reqlen);
+      memcpy(pkt, &mark, sizeof(mark));
     }
 
-  /* Send the call message
+  /* Send a RPC message
+   * For TCP, the record marking + call message will be sent
+   * For UDP, call message will be sent
    *
    * On success, psock_send returns the number of bytes sent;
    * On failure, it returns a negated errno value.
    */
 
-  ret = psock_send(&rpc->rc_so, call, reqlen, 0);
+  if (pkt)
+    {
+      memcpy(pkt + sizeof(mark), call, reqlen);
+      ret = psock_send(&rpc->rc_so, pkt, sizeof(mark) + reqlen, 0);
```

Comment:
it's better to call psock_sendmsg instead to avoid kmalloc in every transaction.

Reply:
@xiaoxiang781216 

I thought the same thing but I found that we need to modify the existing data structure which might give an impact.
Anyway, I will try to investigate again.

Reply:
@xiaoxiang781216 
Done.